### PR TITLE
Add and configure Vagrant for RIOT-OS

### PR DESCRIPTION
**Pull request by cgundogan**
Resulting from an offline discussion with @OlegHahm regarding virtual machines and possibly using vagrant, I compiled this PR that adds a Vagrantfile driving an ubuntu/trusty virtualbox image.

To use this image it is necessary to install vagrant [1], virtualbox **and** the extension pack either from [2] or if available from your package manager.

To start the image, call `vagrant up` from the RIOT root directory.
This will download (only once) the ubuntu/trusty image and then installs all toolchains, the cli-tools
and openocd (0.9.0). This process can take several minutes especially for the first-time call.
If you have more than one network device then you will be prompted to choose the correct one for bridging to have internet connectivity from within the virtual machine.

Calling `vagrant ssh` will enter the virtual machine. In the home directory is a symbolic link `RIOT-OS`, which is a synchronized directory. Thus, every change that will be made to this directory either from the host system or from the guest system will be synchronized immediately **(don't delete anything)**.
This means: It is still possible to develop for RIOT with your favorite editor in your host system and just use
the vagrant machine for building, flashing and running native.

For now, I included just a couple of udev rules (st-link, ftdi, usbasp). Feel free to add more udev rules, but keep in mind that the Vagrantfile has to be extended to include the right usbfilters (exposing the usb devices to the guest system).

`vagrant halt` powers off the virtual machine gracefully.
`vagrant destroy` deletes all changes to the image and resets it to the default state.
`vagrant up` starts the virtual machine and (if not done before, or if destroyed before) installs all toolchains etc.
`vagrant ssh` logs in to the guest system.

I only have a `stm32f4discovery` board at hand right now and I was able to compile the `default` example and flash the board from within the virtual machine.

@emmanuelsearch @thomaseichinger can you try if running the machine and flashing boards do also work on OS X?

Can anybody try with Windows?

[1] https://www.vagrantup.com/
[2] https://www.virtualbox.org/wiki/Downloads

### CODE REVIEW

**Comment by OlegHahm** [comment]
@cgundogan, can you add a readme, please? Maybe just copying the description above?

@waehlisch, could you test on Windows?

**Comment by cgundogan** [comment]
@OlegHahm I added some info to the existing `README.md` file in the root directory. Tell me if you would rather have a separate `README` file in `dist/tools/vagrant/`.

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  #   # Display the VirtualBox GUI when booting the machine
+  #   vb.gui = true
+  #
+  #   # Customize the amount of memory on the VM:
+  #   vb.memory = "1024"
+    # enable usb passthrough
+    vb.customize ["modifyvm", :id, "--usb", "on"]
+    vb.customize ["modifyvm", :id, "--usbehci", "on"]
+    vb.customize ['usbfilter', 'add', '0', '--target', :id, '--name', 'FTDI USB-TTL',
+                  '--vendorid', '0x0403', '--productid',  '0x6001']
```
Comment (thomaseichinger) [inline Vagrantfile]: iotlab-m3 has product id `0x6010`

File: dist/tools/vagrant/udev_rules/53-usbttl.rules [hunk 2]
```diff
@@ -0,0 +1 @@
+SUBSYSTEM=="usb", ATTR{idVendor}=="0403", ATTR{idProduct}=="6001", MODE="0666", OWNER="vagrant"
```
Comment (thomaseichinger) [inline dist/tools/vagrant/udev_rules/53-usbttl.rules]: s/6001/6010/

**Comment by thomaseichinger** [comment]
Please add a udev rule for iotlab-m3:
```
SUBSYSTEM=="usb", ATTR{idVendor}=="0403", ATTR{idProduct}=="6010", MODE="0666", OWNER="vagrant"
```

**Comment by thomaseichinger** [comment]
Tested for OS X and seems to work. You have to install the VirtualBox Extension back though. Else `vagrant up` will fail without meaningful error message.

**Comment by thomaseichinger** [comment]
With the iotlab-m3 rule flashing also works. Nice job!

File: Vagrantfile (hunk 1 above)
Comment (cgundogan) [inline Vagrantfile]: addressed

**Comment by OlegHahm** [comment]
Just talked with @emmanuelsearch and we both agree that we would prefer a separate README with a link from the Wiki.

**Comment by OlegHahm** [comment]
I get
```
Failed to open a session for the virtual machine RIOT_default_1448964176208_75289.

Failed to open/create the internal network 'HostInterfaceNetworking-eno1' (VERR_SUPDRV_COMPONENT_NOT_FOUND).

Failed to attach the network LUN (VERR_SUPDRV_COMPONENT_NOT_FOUND).

One of the kernel modules was not successfully loaded. Make sure that no kernel modules from an older version of VirtualBox exist. Then try to recompile and reload the kernel modules by executing '/sbin/vboxconfig' as root (VERR_SUPDRV_COMPONENT_NOT_FOUND).

Result Code: NS_ERROR_FAILURE (0x80004005)
Component: ConsoleWrap
Interface: IConsole {872da645-4a9b-1727-bee2-5585105b9eed}
```
when I try to start the VM. Other VMs seem to work. Any idea?

**Comment by cgundogan** [comment]
>...when I try to start the VM. Other VMs seem to work. Any idea?

BTW, this issue was resolved in an offline discussion by installing the virtualbox guest-iso in arch linux

**Comment by OlegHahm** [comment]
I'm happy with the current sate. ACK from my side.

**Comment by OlegHahm** [comment]
(Apart from the README, of course.)

**Comment by cgundogan** [comment]
I am currently facing an issue with the samr21-xpro board (make term does not show any output). Furthermore, I have to look more on the integration with windows as it seems..

>Just talked with @emmanuelsearch and we both agree that we would prefer a separate README with a link from the Wiki.

I will move the info to a separate README within `dist/tools/vagrant/`

**Comment by mziegert** [comment]
Tested this on windows 7 32-bit. (Changed trusty64 => trusty32) and removed the file provisioner line from the Vagrantfile. I was able to flash f4vi1 and samr21-xpro boards. ACK

**Comment by OlegHahm** [comment]
@cgundogan, I guess you can squash then. @mziegert, thanks for testing on Windows.

**Comment by cgundogan** [comment]
squashed

**Comment by emmanuelsearch** [comment]
ACK. Let's merge and fix the ATSAMR21 issue later.

**Comment by emmanuelsearch** [comment]
@cgundogan can you open an issue about the ATSAMR21?
(I suppose you agree with merging now, right? Else let us know what you plan to do ;)

**Comment by cgundogan** [comment]
I would vote for waiting one more day. I did included a new commit that changes the ubuntu image to a more recent ubuntu version. This fixes at least the virtual box guest iso version mismatch. Could you @emmanuelsearch or @thomaseichinger test flashing the sar21-xpro with this current state of this PR on your OS X machines?

**Comment by emmanuelsearch** [comment]
OK will test now.

**Comment by emmanuelsearch** [comment]
Works for me on OS X flashing and term on ATSAMR21. Great! re-ACK ;)

**Comment by emmanuelsearch** [comment]
Yay! this is a big step forward for average OS X users like myself ;) and for Windows users, and for beginners in general!

**Comment by cgundogan** [comment]
travis is green - GO

**Pull request merged**